### PR TITLE
UPSTREAM: 33319: add nodeport option when creating NodePort service

### PR DESCRIPTION
**Pull request by juanvallejo**
UPSTREAM: https://github.com/kubernetes/kubernetes/pull/33319
Related bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1378012

This patch adds the option to set a nodeport when creating a NodePort
service. In case of a port allocation error due to a specified port
being out of the valid range, the error now includes the valid
range. If a `--nodeport` value is not specified, it defaults to zero, in
which case the allocator will default to its current behavior of
assigning an available port.

This patch also adds a new helper function in `cmd/util/helpers.go` to
retrieve `Int32` cobra flags.

**Example**
```
$ oc create service nodeport mynodeport --tcp=8080:7777 --nodeport=1
The Service "mynodeport" is invalid: spec.ports[0].nodePort: Invalid
value: 1: provided port is not in the valid range. Valid ports range
from 30000-32767

$ oc create service nodeport mynodeport --tcp=8080:7777 --nodeport=30000
service "mynodeport" created

$ oc describe service mynodeport
Name:                   mynodeport
Namespace:              default
Labels:                 app=mynodeport
Selector:               app=mynodeport
Type:                   NodePort
IP:                     172.30.81.254
Port:                   8080-7777       8080/TCP
NodePort:               8080-7777       30000/TCP
Endpoints:              <none>
Session Affinity:       None
No events.
```

cc @openshift/cli-review @csrwng

### CODE REVIEW

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/helpers.go [hunk 1]
```diff
@@ -344,6 +344,15 @@ func GetFlagInt64(cmd *cobra.Command, flag string) int64 {
 	return i
 }
 
+// Assumes the flag has a default value.
+func GetFlagInt32(cmd *cobra.Command, flag string) int32 {
+	i, err := cmd.Flags().GetInt32(flag)
+	if err != nil {
+		glog.Fatalf("err accessing flag %s for command %s: %v", flag, cmd.Name(), err)
```
Comment (csrwng) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/util/helpers.go]: s/err/error/

File: vendor/k8s.io/kubernetes/pkg/kubectl/cmd/create_service.go [hunk 2]
```diff
@@ -133,6 +133,7 @@ func NewCmdCreateServiceNodePort(f *cmdutil.Factory, cmdOut io.Writer) *cobra.Co
 	cmdutil.AddValidateFlags(cmd)
 	cmdutil.AddPrinterFlags(cmd)
 	cmdutil.AddGeneratorFlags(cmd, cmdutil.ServiceNodePortGeneratorV1Name)
+	cmd.Flags().Int32("nodeport", 0, "Port used to expose the service on each node in a cluster.")
```
Comment (csrwng) [inline vendor/k8s.io/kubernetes/pkg/kubectl/cmd/create_service.go]: Seems like it should be named "node-port" based on the capitalization used in the API (NodePort)

File: vendor/k8s.io/kubernetes/pkg/registry/service/portallocator/allocator.go [hunk 3]
```diff
@@ -82,7 +82,10 @@ func (r *PortAllocator) Free() int {
 func (r *PortAllocator) Allocate(port int) error {
 	ok, offset := r.contains(port)
 	if !ok {
-		return ErrNotInRange
+		// include valid port range in error
+		validPorts := r.portRange.String()
+		msg := fmt.Sprintf("Valid ports range from %s", validPorts)
```
Comment (csrwng) [inline vendor/k8s.io/kubernetes/pkg/registry/service/portallocator/allocator.go]: nit ... I think "Valid ports range is %s" would be better

**Comment by juanvallejo** [comment]
[test]

**Comment by juanvallejo** [comment]
@csrwng Review comments addressed, thanks for the feedback!

**Comment by juanvallejo** [comment]
[conformance](https://ci.openshift.redhat.com/jenkins/job/test_pull_requests_origin_conformance/6414/) test flaked on https://github.com/openshift/origin/issues/11079 re[test]

**Comment by juanvallejo** [comment]
@fabianofranz or @csrwng [upstream](https://github.com/kubernetes/kubernetes/pull/33319) PR merged, PTAL

**Comment by fabianofranz** [comment]
@juanvallejo is this up-to-date with upstream (4 versus 2 commits)?

**Comment by juanvallejo** [comment]
@fabianofranz it's up to date, I had forgotten to squash upstream :/

**Comment by fabianofranz** [comment]
[merge]

**Pull request merged**